File: 153.find-minimum-in-rotated-sorted-array.go
func findMin(nums []int) int {
    if nums[0] < nums[len(nums)-1] {
        return nums[0]
    }
    return find(nums, 0, len(nums)-1)
}

func find(nums []int, lo, hi int) int {
	if lo == hi {
		return nums[lo]
	}
	mid := (hi + lo) / 2
	if nums[mid] > nums[mid+1] {
		return nums[mid+1]
	}
    if nums[mid] < nums[lo] {
        return find(nums, lo, mid)
    }
    return find(nums, mid+1, hi)
}

File: 153.find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        // 3 4 5 1 2
        // nums[i] <= 2
        // 0 0 0 1 1
        //
        // 0 1 2 3 4
        // nums[i] <= 4
        // 1 1 1 1 1
        int l = 0, r = nums.length - 1;
        int lastVal = nums[r];
        int ans = nums.length;
        while (l <= r) {
            int mid = (int) Math.floor((l + r) / 2);
            int val = this.getVal(nums, mid, lastVal);
            if (val == 1) {
                ans = Math.min(ans, mid);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return nums[ans];
    }

    public int getVal(int[] nums, int idx, int val) {
        return nums[idx] <= val ? 1 : 0;
    }
}
